File: services/function/function_sca_go/pkg/schemas/schemas.go
package schemas

import "time"

// CommonVulnerability 統一的漏洞結構定義
type CommonVulnerability struct {
	// 基本識別信息
	ID          string `json:"id"`          // 漏洞 ID (CVE-2023-1234, GHSA-xxxx-yyyy-zzzz)
	Type        string `json:"type"`        // 漏洞類型 (SCA, SAST, DAST, etc.)
	Name        string `json:"name"`        // 漏洞名稱
	Summary     string `json:"summary"`     // 簡要說明
	Description string `json:"description"` // 詳細描述

	// 嚴重性和評分
	Severity   string  `json:"severity"`              // 嚴重性 (CRITICAL, HIGH, MEDIUM, LOW)
	CVSS       float64 `json:"cvss,omitempty"`        // CVSS 分數
	CVSSVector string  `json:"cvss_vector,omitempty"` // CVSS 向量字串

	// 標準化識別碼
	CVEID  string   `json:"cve_id,omitempty"`  // CVE 編號
	GHSAID string   `json:"ghsa_id,omitempty"` // GitHub Security Advisory ID
	CWEIDs []string `json:"cwe_ids,omitempty"` // CWE 弱點分類
	OSVIDs []string `json:"osv_ids,omitempty"` // OSV 數據庫 ID

	// 影響範圍
	Affected   []string `json:"affected,omitempty"`   // 受影響的版本範圍
	Fixed      []string `json:"fixed,omitempty"`      // 已修復的版本
	Introduced []string `json:"introduced,omitempty"` // 引入漏洞的版本

	// 參考資料
	References []string `json:"references,omitempty"` // 參考連結
	Aliases    []string `json:"aliases,omitempty"`    // 其他別名

	// 時間戳記
	PublishedAt *time.Time `json:"published_at,omitempty"` // 公開日期
	ModifiedAt  *time.Time `json:"modified_at,omitempty"`  // 修改日期

	// 擴展元數據
	Metadata map[string]interface{} `json:"metadata,omitempty"` // 擴展資料
}

// CommonDependency 統一的依賴項結構定義
type CommonDependency struct {
	// 基本信息
	Name      string `json:"name"`      // 套件名稱
	Version   string `json:"version"`   // 版本號
	Language  string `json:"language"`  // 程式語言
	Ecosystem string `json:"ecosystem"` // 生態系統 (npm, PyPI, Maven, etc.)

	// 來源信息
	FilePath   string `json:"file_path"`   // 來源文件路徑
	LineNumber int    `json:"line_number"` // 行號

	// 依賴類型
	Type  string `json:"type"`            // 依賴類型 (direct, indirect, dev)
	Scope string `json:"scope,omitempty"` // 依賴範圍 (compile, runtime, test, etc.)

	// 漏洞信息
	Vulnerabilities []CommonVulnerability `json:"vulnerabilities,omitempty"`

	// 許可證信息
	License    string `json:"license,omitempty"`     // 許可證
	LicenseURL string `json:"license_url,omitempty"` // 許可證連結

	// 元數據
	Metadata map[string]string `json:"metadata,omitempty"`

	// 時間戳記
	CreatedAt *time.Time `json:"created_at,omitempty"`
	UpdatedAt *time.Time `json:"updated_at,omitempty"`
}

// ScanResult 掃描結果統一結構
type ScanResult struct {
	// 掃描基本信息
	ScanID   string `json:"scan_id"`   // 掃描 ID
	TaskID   string `json:"task_id"`   // 任務 ID
	ScanType string `json:"scan_type"` // 掃描類型 (SCA, SAST, DAST)

	// 目標信息
	Target ScanTarget `json:"target"` // 掃描目標

	// 掃描狀態
	Status   string  `json:"status"`   // 狀態 (running, completed, failed)
	Progress float64 `json:"progress"` // 進度 (0-100)

	// 結果統計
	Summary ScanSummary `json:"summary"` // 結果摘要

	// 詳細結果
	Dependencies []CommonDependency `json:"dependencies,omitempty"` // 發現的依賴
	Findings     []FindingPayload   `json:"findings,omitempty"`     // 發現的問題

	// 時間信息
	StartedAt   time.Time  `json:"started_at"`             // 開始時間
	CompletedAt *time.Time `json:"completed_at,omitempty"` // 完成時間
	Duration    int64      `json:"duration,omitempty"`     // 掃描時長(秒)

	// 配置和元數據
	Config   map[string]interface{} `json:"config,omitempty"`   // 掃描配置
	Metadata map[string]interface{} `json:"metadata,omitempty"` // 擴展元數據

	// 錯誤信息
	Error *ScanError `json:"error,omitempty"` // 錯誤詳情
}

// ScanTarget 掃描目標
type ScanTarget struct {
	Type       string            `json:"type"`                 // 目標類型 (repository, file, url)
	URL        string            `json:"url"`                  // 目標 URL
	Path       string            `json:"path,omitempty"`       // 本地路徑
	Branch     string            `json:"branch,omitempty"`     // Git 分支
	Commit     string            `json:"commit,omitempty"`     // Git 提交
	Method     string            `json:"method,omitempty"`     // HTTP 方法
	Headers    map[string]string `json:"headers,omitempty"`    // HTTP 標頭
	Parameters map[string]string `json:"parameters,omitempty"` // 參數
}

// ScanSummary 掃描結果摘要
type ScanSummary struct {
	// 統計數字
	TotalDependencies int `json:"total_dependencies"` // 總依賴數
	TotalFindings     int `json:"total_findings"`     // 總發現數

	// 按嚴重性分組
	Critical int `json:"critical"` // 嚴重漏洞數
	High     int `json:"high"`     // 高危漏洞數
	Medium   int `json:"medium"`   // 中危漏洞數
	Low      int `json:"low"`      // 低危漏洞數
	Info     int `json:"info"`     // 信息級別數

	// 按類型分組
	ByType map[string]int `json:"by_type,omitempty"` // 按漏洞類型分組

	// 按語言分組
	ByLanguage map[string]int `json:"by_language,omitempty"` // 按程式語言分組
}

// ScanError 掃描錯誤信息
type ScanError struct {
	Code        string    `json:"code"`              // 錯誤代碼
	Message     string    `json:"message"`           // 錯誤訊息
	Details     string    `json:"details,omitempty"` // 錯誤詳情
	Timestamp   time.Time `json:"timestamp"`         // 發生時間
	Recoverable bool      `json:"recoverable"`       // 是否可恢復
}

// FindingPayload 統一的發現載荷結構
type FindingPayload struct {
	// 基本識別
	FindingID string `json:"finding_id"`        // 發現 ID
	TaskID    string `json:"task_id"`           // 任務 ID
	ScanID    string `json:"scan_id,omitempty"` // 掃描 ID

	// 漏洞信息
	Vulnerability CommonVulnerability `json:"vulnerability"` // 漏洞詳情

	// 嚴重性和可信度
	Severity   string `json:"severity"`   // 嚴重性
	Confidence string `json:"confidence"` // 可信度 (CERTAIN, FIRM, TENTATIVE)

	// 目標和位置
	Target   FindingTarget   `json:"target"`             // 發現目標
	Location FindingLocation `json:"location,omitempty"` // 具體位置

	// 證據和影響
	Evidence       FindingEvidence       `json:"evidence"`       // 證據
	Impact         FindingImpact         `json:"impact"`         // 影響
	Recommendation FindingRecommendation `json:"recommendation"` // 建議

	// 分類和標籤
	Category string   `json:"category,omitempty"` // 分類
	Tags     []string `json:"tags,omitempty"`     // 標籤

	// 狀態管理
	Status     string     `json:"status"`                // 狀態 (new, confirmed, false_positive, fixed)
	AssignedTo string     `json:"assigned_to,omitempty"` // 負責人
	CreatedAt  time.Time  `json:"created_at"`            // 創建時間
	UpdatedAt  time.Time  `json:"updated_at"`            // 更新時間
	ResolvedAt *time.Time `json:"resolved_at,omitempty"` // 解決時間

	// 擴展信息
	Metadata map[string]interface{} `json:"metadata,omitempty"` // 元數據
}

// FindingTarget 發現目標 (與 models 保持一致)
type FindingTarget struct {
	URL       string `json:"url"`
	Method    string `json:"method,omitempty"`
	Parameter string `json:"parameter,omitempty"`
}

// FindingLocation 發現的具體位置
type FindingLocation struct {
	FilePath    string `json:"file_path,omitempty"`    // 文件路徑
	LineStart   int    `json:"line_start,omitempty"`   // 起始行
	LineEnd     int    `json:"line_end,omitempty"`     // 結束行
	ColumnStart int    `json:"column_start,omitempty"` // 起始列
	ColumnEnd   int    `json:"column_end,omitempty"`   // 結束列
	Function    string `json:"function,omitempty"`     // 函數名
	Class       string `json:"class,omitempty"`        // 類名
}

// FindingEvidence 證據 (與 models 保持一致)
type FindingEvidence struct {
	Request        string `json:"request"`
	Response       string `json:"response"`
	Payload        string `json:"payload,omitempty"`
	ProofOfConcept string `json:"proof_of_concept,omitempty"`
}

// FindingImpact 影響 (與 models 保持一致)
type FindingImpact struct {
	Confidentiality string `json:"confidentiality"`
	Integrity       string `json:"integrity"`
	Availability    string `json:"availability"`
	BusinessImpact  string `json:"business_impact"`
}

// FindingRecommendation 修復建議 (與 models 保持一致)
type FindingRecommendation struct {
	Remediation string   `json:"remediation"`
	References  []string `json:"references"`
}

File: plugins/cli_generated/schemas.ts
/**
 * AIVA Common TypeScript Schemas - 自動生成
 * ==========================================
 * 
 * AIVA跨語言Schema統一定義 - 以手動維護版本為準
 * 
 * ⚠️  此配置已同步手動維護的Schema定義，確保單一事實原則
 * 📅 最後更新: 2025-10-30T00:00:00.000000
 * 🔄 Schema 版本: 1.1.0
 * 
 * 遵循單一事實原則，與 Python aiva_common.schemas 保持完全一致
 */

// ==================== 枚舉類型定義 ====================

export enum Severity {
  CRITICAL = "critical",
  HIGH = "high",
  MEDIUM = "medium",
  LOW = "low",
  INFORMATIONAL = "info"
}

export enum Confidence {
  CERTAIN = "certain",
  FIRM = "firm",
  POSSIBLE = "possible"
}

export enum VulnerabilityType {
  XSS = "XSS",
  SQLI = "SQL Injection",
  SSRF = "SSRF",
  IDOR = "IDOR",
  BOLA = "BOLA",
  INFO_LEAK = "Information Leak",
  WEAK_AUTH = "Weak Authentication",
  RCE = "Remote Code Execution",
  AUTHENTICATION_BYPASS = "Authentication Bypass"
}

// ==================== 基礎介面定義 ====================

/**
 * 訊息標頭 - 用於所有訊息的統一標頭格式
 */
export interface MessageHeader {
  message_id: string;
  trace_id: string;
  correlation_id?: string | null;
  /** 來源模組名稱 */
  source_module: string;
  timestamp?: string;
  version?: string;
}

/**
 * 目標資訊 - 漏洞所在位置
 */
export interface Target {
  url: any;
  parameter?: string | null;
  method?: string | null;
  headers?: Record<string, any>;
  params?: Record<string, any>;
  body?: string | null;
}

/**
 * 漏洞基本資訊 - 用於 Finding 中的漏洞描述。符合標準：CWE、CVE、CVSS v3.1/v4.0、OWASP
 */
export interface Vulnerability {
  name: any;
  /** CWE ID (格式: CWE-XXX)，參考 https://cwe.mitre.org/ */
  cwe?: string | null;
  /** CVE ID (格式: CVE-YYYY-NNNNN)，參考 https://cve.mitre.org/ */
  cve?: string | null;
  severity: any;
  confidence: any;
  description?: string | null;
  /** CVSS v3.1 Base Score (0.0-10.0)，參考 https://www.first.org/cvss/ */
  cvss_score?: any;
  /** CVSS v3.1 Vector String，例如: CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H */
  cvss_vector?: string | null;
  /** OWASP Top 10 分類，例如: A03:2021-Injection */
  owasp_category?: string | null;
}

/**
 * 資產基本資訊
 */
export interface Asset {
  asset_id: string;
  type: string;
  value: string;
  parameters?: string[];
  has_form?: boolean;
}

/**
 * 認證資訊
 */
export interface Authentication {
  method?: string;
  credentials?: Record<string, any>;
}

/**
 * 執行錯誤統一格式
 */
export interface ExecutionError {
  error_id: string;
  error_type: string;
  message: string;
  payload?: string | null;
  vector?: string | null;
  timestamp?: string;
  attempts?: number;
}

/**
 * 技術指紋
 */
export interface Fingerprints {
  web_server?: Record<string, any>;
  framework?: Record<string, any>;
  language?: Record<string, any>;
  waf_detected?: boolean;
  waf_vendor?: string | null;
}

/**
 * 速率限制
 */
export interface RateLimit {
  requests_per_second?: number;
  burst?: number;
}

/**
 * 風險因子
 */
export interface RiskFactor {
  /** 風險因子名稱 */
  factor_name: string;
  /** 權重 */
  weight: number;
  /** 因子值 */
  value: number;
  /** 因子描述 */
  description?: string | null;
}

/**
 * 掃描範圍
 */
export interface ScanScope {
  exclusions?: string[];
  include_subdomains?: boolean;
  allowed_hosts?: string[];
}

/**
 * 掃描摘要
 */
export interface Summary {
  urls_found?: number;
  forms_found?: number;
  apis_found?: number;
  scan_duration_seconds?: number;
}

/**
 * 任務依賴
 */
export interface TaskDependency {
  /** 依賴類型 */
  dependency_type: string;
  /** 依賴任務ID */
  dependent_task_id: string;
  /** 依賴條件 */
  condition?: string | null;
  /** 是否必需 */
  required?: boolean;
}

/**
 * AI 驅動漏洞驗證請求
 */
export interface AIVerificationRequest {
  verification_id: string;
  finding_id: string;
  scan_id: string;
  vulnerability_type: any;
  target: any;
  evidence: any;
  verification_mode?: string;
  context?: Record<string, any>;
}

/**
 * AI 驅動漏洞驗證結果
 */
export interface AIVerificationResult {
  verification_id: string;
  finding_id: string;
  verification_status: string;
  confidence_score: number;
  verification_method: string;
  test_steps?: string[];
  observations?: string[];
  recommendations?: string[];
  timestamp?: string;
}

/**
 * 程式碼層面根因分析結果
 */
export interface CodeLevelRootCause {
  analysis_id: string;
  vulnerable_component: string;
  affected_findings: string[];
  code_location?: string | null;
  vulnerability_pattern?: string | null;
  fix_recommendation?: string | null;
}

/**
 * 目標資訊 - 漏洞所在位置
 */
export interface FindingTarget {
  url: any;
  parameter?: string | null;
  method?: string | null;
  headers?: Record<string, any>;
  params?: Record<string, any>;
  body?: string | null;
}

/**
 * JavaScript 分析結果
 */
export interface JavaScriptAnalysisResult {
  analysis_id: string;
  url: string;
  source_size_bytes: number;
  dangerous_functions?: string[];
  external_resources?: string[];
  data_leaks?: Record<string, any>;
  findings?: string[];
  apis_called?: string[];
  ajax_endpoints?: string[];
  suspicious_patterns?: string[];
  risk_score?: number;
  security_score?: number;
  timestamp?: string;
}

/**
 * SAST-DAST 資料流關聯結果
 */
export interface SASTDASTCorrelation {
  correlation_id: string;
  sast_finding_id: string;
  dast_finding_id: string;
  data_flow_path: string[];
  verification_status: string;
  confidence_score: number;
  explanation?: string | null;
}

/**
 * 敏感資訊匹配結果
 */
export interface SensitiveMatch {
  match_id: string;
  pattern_name: string;
  matched_text: string;
  context: string;
  confidence: number;
  line_number?: any;
  file_path?: string | null;
  url?: string | null;
  severity?: any;
}

/**
 * 漏洞關聯分析結果
 */
export interface VulnerabilityCorrelation {
  correlation_id: string;
  correlation_type: string;
  related_findings: string[];
  confidence_score: number;
  root_cause?: string | null;
  common_components?: string[];
  explanation?: string | null;
  timestamp?: string;
}

// ==================== 訊息通訊類型 ====================

/**
 * AIVA統一訊息格式 - 所有跨服務通訊的標準信封
 */
export interface AivaMessage {
  /** 訊息標頭 */
  header: MessageHeader;
  /** 訊息主題 */
  topic: string;
  /** Schema版本 */
  schema_version: string;
  /** 訊息載荷 */
  payload: Record<string, any>;
}

/**
 * 統一請求格式 - 模組間請求通訊
 */
export interface AIVARequest {
  /** 請求識別碼 */
  request_id: string;
  /** 來源模組 */
  source_module: string;
  /** 目標模組 */
  target_module: string;
  /** 請求類型 */
  request_type: string;
  /** 請求載荷 */
  payload: Record<string, any>;
  /** 追蹤識別碼 */
  trace_id?: string | null;
  /** 逾時秒數 */
  timeout_seconds: number;
  /** 中繼資料 */
  metadata?: Record<string, any>;
  /** 時間戳 */
  timestamp: string;
}

/**
 * 統一響應格式 - 模組間響應通訊
 */
export interface AIVAResponse {
  /** 對應的請求識別碼 */
  request_id: string;
  /** 響應類型 */
  response_type: string;
  /** 執行是否成功 */
  success: boolean;
  /** 響應載荷 */
  payload?: Record<string, any> | null;
  /** 錯誤代碼 */
  error_code?: string | null;
  /** 錯誤訊息 */
  error_message?: string | null;
  /** 中繼資料 */
  metadata?: Record<string, any>;
  /** 時間戳 */
  timestamp: string;
}

// ==================== 漏洞發現類型 ====================

/**
 * 漏洞發現載荷 - 掃描結果的標準格式
 */
export interface FindingPayload {
  /** 發現識別碼 */
  finding_id: string;
  /** 任務識別碼 */
  task_id: string;
  /** 掃描識別碼 */
  scan_id: string;
  /** 發現狀態 */
  status: string;
  /** 漏洞資訊 */
  vulnerability: Vulnerability;
  /** 目標資訊 */
  target: Target;
  /** 使用的策略 */
  strategy?: string | null;
  /** 證據資料 */
  evidence?: FindingEvidence | null;
  /** 影響評估 */
  impact?: FindingImpact | null;
  /** 修復建議 */
  recommendation?: FindingRecommendation | null;
  /** 中繼資料 */
  metadata?: Record<string, any>;
  /** 建立時間 */
  created_at: string;
  /** 更新時間 */
  updated_at: string;
}

/**
 * 漏洞證據
 */
export interface FindingEvidence {
  /** 攻擊載荷 */
  payload?: string | null;
  /** 響應時間差異 */
  response_time_delta?: number | null;
  /** 資料庫版本 */
  db_version?: string | null;
  /** HTTP請求 */
  request?: string | null;
  /** HTTP響應 */
  response?: string | null;
  /** 證明資料 */
  proof?: string | null;
}

/**
 * 漏洞影響評估
 */
export interface FindingImpact {
  /** 影響描述 */
  description?: string | null;
  /** 業務影響 */
  business_impact?: string | null;
  /** 技術影響 */
  technical_impact?: string | null;
  /** 受影響用戶數 */
  affected_users?: number | null;
  /** 估計成本 */
  estimated_cost?: number | null;
}

/**
 * 漏洞修復建議
 */
export interface FindingRecommendation {
  /** 修復方法 */
  fix?: string | null;
  /** 修復優先級 */
  priority?: string | null;
  /** 修復步驟 */
  remediation_steps?: string[];
  /** 參考資料 */
  references?: string[];
}

/**
 * Token 測試結果
 */
export interface TokenTestResult {
  /** 是否存在漏洞 */
  vulnerable: boolean;
  /** Token 類型 (jwt, session, api, etc.) */
  token_type: string;
  /** 發現的問題 */
  issue: string;
  /** 詳細描述 */
  details: string;
  /** 解碼後的載荷內容 */
  decoded_payload?: Record<string, any> | null;
  /** 漏洞嚴重程度 */
  severity?: string;
  /** 測試類型 */
  test_type: string;
}

// ==================== 任務管理類型 ====================

/**
 * 功能任務載荷 - 掃描任務的標準格式
 */
export interface FunctionTaskPayload {
  /** 任務識別碼 */
  task_id: string;
  /** 掃描識別碼 */
  scan_id: string;
  /** 任務優先級 */
  priority: number;
  /** 掃描目標 */
  target: FunctionTaskTarget;
  /** 任務上下文 */
  context: FunctionTaskContext;
  /** 掃描策略 */
  strategy: string;
  /** 自訂載荷 */
  custom_payloads?: string[];
  /** 測試配置 */
  test_config: FunctionTaskTestConfig;
}

/**
 * 功能任務目標
 */
export interface FunctionTaskTarget {
}

/**
 * 功能任務上下文
 */
export interface FunctionTaskContext {
  /** 資料庫類型提示 */
  db_type_hint?: string | null;
  /** 是否檢測到WAF */
  waf_detected: boolean;
  /** 相關發現 */
  related_findings?: string[];
}

/**
 * 功能任務測試配置
 */
export interface FunctionTaskTestConfig {
  /** 標準載荷列表 */
  payloads: string[];
  /** 自訂載荷列表 */
  custom_payloads?: string[];
  /** 是否進行Blind XSS測試 */
  blind_xss: boolean;
  /** 是否進行DOM測試 */
  dom_testing: boolean;
  /** 請求逾時(秒) */
  timeout?: number | null;
}

/**
 * 掃描任務載荷 - 用於SCA/SAST等需要項目URL的掃描任務
 */
export interface ScanTaskPayload {
  /** 任務識別碼 */
  task_id: string;
  /** 掃描識別碼 */
  scan_id: string;
  /** 任務優先級 */
  priority: number;
  /** 掃描目標 (包含URL) */
  target: Target;
  /** 掃描類型 */
  scan_type: string;
  /** 代碼倉庫資訊 (分支、commit等) */
  repository_info?: Record<string, any> | null;
  /** 掃描逾時(秒) */
  timeout?: number | null;
}
